stackslib/src/core/mempool.rs

```diff
@@ -1660,172 +1617,235 @@ impl MemPoolDB {
+            //    by a random number. This helps us mix these transactions with others to guarantee they get processed in a reasonable
+            //    order
+            // 3. Ranks transactions by prioritizing those with next nonces and higher fees (per origin and sponsor address)
+            // 4. Takes the top ranked transaction and returns it for evaluation
+            //
+            // This logic prevents miners from repeatedly visiting (and then skipping) high fee transactions that would get evaluated
+            // first based on their `fee_rate` but are otherwise non-mineable because they have very high or invalid nonces. A large
+            // volume of these transactions would cause considerable slowness when selecting valid transactions to mine. This query
+            // also makes sure transactions that have NULL `fee_rate`s are visited, because they will also get ranked according to
+            // their origin address nonce.
+            let sql = "
+            WITH nonce_filtered AS (
+                SELECT txid, origin_nonce, origin_address, sponsor_nonce, sponsor_address, fee_rate,
+                    CASE
+                        WHEN fee_rate IS NULL THEN (ABS(RANDOM()) % 10000 / 10000.0) * (SELECT MAX(fee_rate) FROM mempool)
+                        ELSE fee_rate
+                    END AS sort_fee_rate
+                FROM mempool AS m
+                LEFT JOIN nonces AS no ON m.origin_address = no.address
+                LEFT JOIN nonces AS ns ON m.sponsor_address = ns.address
+                WHERE (no.address IS NULL OR m.origin_nonce = no.nonce)
+                    AND (ns.address IS NULL OR m.sponsor_nonce = ns.nonce)
+                    AND m.txid NOT IN (SELECT txid FROM considered_txs)
+                ORDER BY accept_time ASC
+                LIMIT 11650 -- max transactions that can fit in one block
+            ),
+            address_nonce_ranked AS (
+                SELECT *,
+                    ROW_NUMBER() OVER (
+                        PARTITION BY origin_address
+                        ORDER BY origin_nonce ASC, sort_fee_rate DESC
+                    ) AS origin_rank,
+                    ROW_NUMBER() OVER (
+                        PARTITION BY sponsor_address
+                        ORDER BY sponsor_nonce ASC, sort_fee_rate DESC
+                    ) AS sponsor_rank
+                FROM nonce_filtered
+            )
+            SELECT txid, origin_nonce, origin_address, sponsor_nonce, sponsor_address, fee_rate
+            FROM address_nonce_ranked
+            ORDER BY origin_rank ASC, sponsor_rank ASC, sort_fee_rate DESC
+            ";
+            let mut query_stmt_nonce_rank = self.db.prepare(sql).map_err(Error::SqliteError)?;
+            let mut nonce_rank_iterator = query_stmt_nonce_rank
+                .query(NO_PARAMS)
+                .map_err(Error::SqliteError)?;
+
+            let stop_reason = loop {
```

Comment:
Can you add some code comment around the two loops, in particular to explain why two loops are necessary.

Reply:
Added in 30a68d8d1c. Let me know if you have any thoughts on improvements.